File: sdk/swift/Sources/Schemas/Message.swift
import Foundation

public struct Message: Codable, Hashable, Sendable {
    public let id: String
    public let inboxId: String
    public let threadId: String
    public let direction: MessageDirection
    public let status: MessageStatus
    public let from: MessageFrom
    public let to: [MessageToItem]
    public let cc: [MessageCcItem]
    public let bcc: [MessageBccItem]
    public let replyTo: [MessageReplyToItem]
    public let subject: String
    public let text: Nullable<String>
    /// Untrusted email HTML. Sanitize and isolate it before browser rendering.
    public let html: Nullable<String>
    public let extractedText: Nullable<String>
    /// Untrusted latest-reply HTML. Sanitize and isolate it before browser rendering.
    public let extractedHtml: Nullable<String>
    public let headers: [String: String]
    public let inReplyTo: Nullable<String>
    public let references: [String]
    public let attachments: [MessageAttachmentsItem]
    public let createdAt: Date
    public let sentAt: Nullable<Date>
    public let receivedAt: Nullable<Date>
    /// Additional properties that are not explicitly defined in the schema
    public let additionalProperties: [String: JSONValue]

    public init(
        id: String,
        inboxId: String,
        threadId: String,
        direction: MessageDirection,
        status: MessageStatus,
        from: MessageFrom,
        to: [MessageToItem],
        cc: [MessageCcItem],
        bcc: [MessageBccItem],
        replyTo: [MessageReplyToItem],
        subject: String,
        text: Nullable<String>,
        html: Nullable<String>,
        extractedText: Nullable<String>,
        extractedHtml: Nullable<String>,
        headers: [String: String],
        inReplyTo: Nullable<String>,
        references: [String],
        attachments: [MessageAttachmentsItem],
        createdAt: Date,
        sentAt: Nullable<Date>,
        receivedAt: Nullable<Date>,
        additionalProperties: [String: JSONValue] = .init()
    ) {
        self.id = id
        self.inboxId = inboxId
        self.threadId = threadId
        self.direction = direction
        self.status = status
        self.from = from
        self.to = to
        self.cc = cc
        self.bcc = bcc
        self.replyTo = replyTo
        self.subject = subject
        self.text = text
        self.html = html
        self.extractedText = extractedText
        self.extractedHtml = extractedHtml
        self.headers = headers
        self.inReplyTo = inReplyTo
        self.references = references
        self.attachments = attachments
        self.createdAt = createdAt
        self.sentAt = sentAt
        self.receivedAt = receivedAt
        self.additionalProperties = additionalProperties
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.id = try container.decode(String.self, forKey: .id)
        self.inboxId = try container.decode(String.self, forKey: .inboxId)
        self.threadId = try container.decode(String.self, forKey: .threadId)
        self.direction = try container.decode(MessageDirection.self, forKey: .direction)
        self.status = try container.decode(MessageStatus.self, forKey: .status)
        self.from = try container.decode(MessageFrom.self, forKey: .from)
        self.to = try container.decode([MessageToItem].self, forKey: .to)
        self.cc = try container.decode([MessageCcItem].self, forKey: .cc)
        self.bcc = try container.decode([MessageBccItem].self, forKey: .bcc)
        self.replyTo = try container.decode([MessageReplyToItem].self, forKey: .replyTo)
        self.subject = try container.decode(String.self, forKey: .subject)
        self.text = try container.decode(Nullable<String>.self, forKey: .text)
        self.html = try container.decode(Nullable<String>.self, forKey: .html)
        self.extractedText = try container.decode(Nullable<String>.self, forKey: .extractedText)
        self.extractedHtml = try container.decode(Nullable<String>.self, forKey: .extractedHtml)
        self.headers = try container.decode([String: String].self, forKey: .headers)
        self.inReplyTo = try container.decode(Nullable<String>.self, forKey: .inReplyTo)
        self.references = try container.decode([String].self, forKey: .references)
        self.attachments = try container.decode([MessageAttachmentsItem].self, forKey: .attachments)
        self.createdAt = try container.decode(Date.self, forKey: .createdAt)
        self.sentAt = try container.decode(Nullable<Date>.self, forKey: .sentAt)
        self.receivedAt = try container.decode(Nullable<Date>.self, forKey: .receivedAt)
        self.additionalProperties = try decoder.decodeAdditionalProperties(using: CodingKeys.self)
    }

    public func encode(to encoder: Encoder) throws -> Void {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try encoder.encodeAdditionalProperties(self.additionalProperties)
        try container.encode(self.id, forKey: .id)
        try container.encode(self.inboxId, forKey: .inboxId)
        try container.encode(self.threadId, forKey: .threadId)
        try container.encode(self.direction, forKey: .direction)
        try container.encode(self.status, forKey: .status)
        try container.encode(self.from, forKey: .from)
        try container.encode(self.to, forKey: .to)
        try container.encode(self.cc, forKey: .cc)
        try container.encode(self.bcc, forKey: .bcc)
        try container.encode(self.replyTo, forKey: .replyTo)
        try container.encode(self.subject, forKey: .subject)
        try container.encode(self.text, forKey: .text)
        try container.encode(self.html, forKey: .html)
        try container.encode(self.extractedText, forKey: .extractedText)
        try container.encode(self.extractedHtml, forKey: .extractedHtml)
        try container.encode(self.headers, forKey: .headers)
        try container.encode(self.inReplyTo, forKey: .inReplyTo)
        try container.encode(self.references, forKey: .references)
        try container.encode(self.attachments, forKey: .attachments)
        try container.encode(self.createdAt, forKey: .createdAt)
        try container.encode(self.sentAt, forKey: .sentAt)
        try container.encode(self.receivedAt, forKey: .receivedAt)
    }

    /// Keys for encoding/decoding struct properties.
    enum CodingKeys: String, CodingKey, CaseIterable {
        case id
        case inboxId
        case threadId
        case direction
        case status
        case from
        case to
        case cc
        case bcc
        case replyTo
        case subject
        case text
        case html
        case extractedText
        case extractedHtml
        case headers
        case inReplyTo
        case references
        case attachments
        case createdAt
        case sentAt
        case receivedAt
    }
}

File: sdk/rust/src/api/types/message.rs
pub use crate::prelude::*;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Message {
    #[serde(default)]
    pub id: String,
    #[serde(rename = "inboxId")]
    #[serde(default)]
    pub inbox_id: String,
    #[serde(rename = "threadId")]
    #[serde(default)]
    pub thread_id: String,
    pub direction: MessageDirection,
    pub status: MessageStatus,
    pub from: MessageFrom,
    #[serde(default)]
    pub to: Vec<MessageToItem>,
    #[serde(default)]
    pub cc: Vec<MessageCcItem>,
    #[serde(default)]
    pub bcc: Vec<MessageBccItem>,
    #[serde(rename = "replyTo")]
    #[serde(default)]
    pub reply_to: Vec<MessageReplyToItem>,
    #[serde(default)]
    pub subject: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    /// Untrusted email HTML. Sanitize and isolate it before browser rendering.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub html: Option<String>,
    #[serde(rename = "extractedText")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extracted_text: Option<String>,
    /// Untrusted latest-reply HTML. Sanitize and isolate it before browser rendering.
    #[serde(rename = "extractedHtml")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extracted_html: Option<String>,
    #[serde(default)]
    pub headers: HashMap<String, String>,
    #[serde(rename = "inReplyTo")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub in_reply_to: Option<String>,
    #[serde(default)]
    pub references: Vec<String>,
    #[serde(default)]
    pub attachments: Vec<MessageAttachmentsItem>,
    #[serde(rename = "createdAt")]
    #[serde(default)]
    #[serde(with = "crate::core::flexible_datetime::offset")]
    pub created_at: DateTime<FixedOffset>,
    #[serde(rename = "sentAt")]
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    #[serde(with = "crate::core::flexible_datetime::offset::option")]
    pub sent_at: Option<DateTime<FixedOffset>>,
    #[serde(rename = "receivedAt")]
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    #[serde(with = "crate::core::flexible_datetime::offset::option")]
    pub received_at: Option<DateTime<FixedOffset>>,
}

impl Message {
    pub fn builder() -> MessageBuilder {
        <MessageBuilder as Default>::default()
    }
}

#[derive(Clone, PartialEq, Default, Debug)]
#[non_exhaustive]
pub struct MessageBuilder {
    id: Option<String>,
    inbox_id: Option<String>,
    thread_id: Option<String>,
    direction: Option<MessageDirection>,
    status: Option<MessageStatus>,
    from: Option<MessageFrom>,
    to: Option<Vec<MessageToItem>>,
    cc: Option<Vec<MessageCcItem>>,
    bcc: Option<Vec<MessageBccItem>>,
    reply_to: Option<Vec<MessageReplyToItem>>,
    subject: Option<String>,
    text: Option<String>,
    html: Option<String>,
    extracted_text: Option<String>,
    extracted_html: Option<String>,
    headers: Option<HashMap<String, String>>,
    in_reply_to: Option<String>,
    references: Option<Vec<String>>,
    attachments: Option<Vec<MessageAttachmentsItem>>,
    created_at: Option<DateTime<FixedOffset>>,
    sent_at: Option<DateTime<FixedOffset>>,
    received_at: Option<DateTime<FixedOffset>>,
}

impl MessageBuilder {
    pub fn id(mut self, value: impl Into<String>) -> Self {
        self.id = Some(value.into());
        self
    }

    pub fn inbox_id(mut self, value: impl Into<String>) -> Self {
        self.inbox_id = Some(value.into());
        self
    }

    pub fn thread_id(mut self, value: impl Into<String>) -> Self {
        self.thread_id = Some(value.into());
        self
    }

    pub fn direction(mut self, value: MessageDirection) -> Self {
        self.direction = Some(value);
        self
    }

    pub fn status(mut self, value: MessageStatus) -> Self {
        self.status = Some(value);
        self
    }

    pub fn from(mut self, value: MessageFrom) -> Self {
        self.from = Some(value);
        self
    }

    pub fn to(mut self, value: Vec<MessageToItem>) -> Self {
        self.to = Some(value);
        self
    }

    pub fn cc(mut self, value: Vec<MessageCcItem>) -> Self {
        self.cc = Some(value);
        self
    }

    pub fn bcc(mut self, value: Vec<MessageBccItem>) -> Self {
        self.bcc = Some(value);
        self
    }

    pub fn reply_to(mut self, value: Vec<MessageReplyToItem>) -> Self {
        self.reply_to = Some(value);
        self
    }

    pub fn subject(mut self, value: impl Into<String>) -> Self {
        self.subject = Some(value.into());
        self
    }

    pub fn text(mut self, value: impl Into<String>) -> Self {
        self.text = Some(value.into());
        self
    }

    pub fn html(mut self, value: impl Into<String>) -> Self {
        self.html = Some(value.into());
        self
    }

    pub fn extracted_text(mut self, value: impl Into<String>) -> Self {
        self.extracted_text = Some(value.into());
        self
    }

    pub fn extracted_html(mut self, value: impl Into<String>) -> Self {
        self.extracted_html = Some(value.into());
        self
    }

    pub fn headers(mut self, value: HashMap<String, String>) -> Self {
        self.headers = Some(value);
        self
    }

    pub fn in_reply_to(mut self, value: impl Into<String>) -> Self {
        self.in_reply_to = Some(value.into());
        self
    }

    pub fn references(mut self, value: Vec<String>) -> Self {
        self.references = Some(value);
        self
    }

    pub fn attachments(mut self, value: Vec<MessageAttachmentsItem>) -> Self {
        self.attachments = Some(value);
        self
    }

    pub fn created_at(mut self, value: DateTime<FixedOffset>) -> Self {
        self.created_at = Some(value);
        self
    }

    pub fn sent_at(mut self, value: DateTime<FixedOffset>) -> Self {
        self.sent_at = Some(value);
        self
    }

    pub fn received_at(mut self, value: DateTime<FixedOffset>) -> Self {
        self.received_at = Some(value);
        self
    }

    /// Consumes the builder and constructs a [`Message`].
    /// This method will fail if any of the following fields are not set:
    /// - [`id`](MessageBuilder::id)
    /// - [`inbox_id`](MessageBuilder::inbox_id)
    /// - [`thread_id`](MessageBuilder::thread_id)
    /// - [`direction`](MessageBuilder::direction)
    /// - [`status`](MessageBuilder::status)
    /// - [`from`](MessageBuilder::from)
    /// - [`to`](MessageBuilder::to)
    /// - [`cc`](MessageBuilder::cc)
    /// - [`bcc`](MessageBuilder::bcc)
    /// - [`reply_to`](MessageBuilder::reply_to)
    /// - [`subject`](MessageBuilder::subject)
    /// - [`headers`](MessageBuilder::headers)
    /// - [`references`](MessageBuilder::references)
    /// - [`attachments`](MessageBuilder::attachments)
    /// - [`created_at`](MessageBuilder::created_at)
    pub fn build(self) -> Result<Message, BuildError> {
        Ok(Message {
            id: self.id.ok_or_else(|| BuildError::missing_field("id"))?,
            inbox_id: self
                .inbox_id
                .ok_or_else(|| BuildError::missing_field("inbox_id"))?,
            thread_id: self
                .thread_id
                .ok_or_else(|| BuildError::missing_field("thread_id"))?,
            direction: self
                .direction
                .ok_or_else(|| BuildError::missing_field("direction"))?,
            status: self
                .status
                .ok_or_else(|| BuildError::missing_field("status"))?,
            from: self.from.ok_or_else(|| BuildError::missing_field("from"))?,
            to: self.to.ok_or_else(|| BuildError::missing_field("to"))?,
            cc: self.cc.ok_or_else(|| BuildError::missing_field("cc"))?,
            bcc: self.bcc.ok_or_else(|| BuildError::missing_field("bcc"))?,
            reply_to: self
                .reply_to
                .ok_or_else(|| BuildError::missing_field("reply_to"))?,
            subject: self
                .subject
                .ok_or_else(|| BuildError::missing_field("subject"))?,
            text: self.text,
            html: self.html,
            extracted_text: self.extracted_text,
            extracted_html: self.extracted_html,
            headers: self
                .headers
                .ok_or_else(|| BuildError::missing_field("headers"))?,
            in_reply_to: self.in_reply_to,
            references: self
                .references
                .ok_or_else(|| BuildError::missing_field("references"))?,
            attachments: self
                .attachments
                .ok_or_else(|| BuildError::missing_field("attachments"))?,
            created_at: self
                .created_at
                .ok_or_else(|| BuildError::missing_field("created_at"))?,
            sent_at: self.sent_at,
            received_at: self.received_at,
        })
    }
}
